gnat2goto/driver/tree_walk.adb

```diff
@@ -2806,6 +2809,204 @@ package body Tree_Walk is
          I_Type => Do_Type_Reference (Etype (N)));
    end Do_Op_Abs;
 
+   function Make_Mod_Expon_Function
+     (Mod_Type : Irep;
+      Exponent_Type : Irep;
+      Source_Location : Source_Ptr)
+   return Irep;
+
+--  'fast' exponentiation, using the fact that
+--  if e is divisible by two, then
+--  b^e == b^(2 * (e/2)) == (b*b)^(e/2)
+--  and otherwise if e == 2p + 1
+--  then b^e = b * b^(2p)
+--  so we only have to do log2(e) multiplications
+--  rather than e
+--
+--  generates a function like this:
+--  modN exp_by_squaring(modN base, unsigned exponent) {
+--    modN exp_result = 1;
+--    while(exponent > 0) {
+--      if(exponent % 2 == 0) {
+--        base *= base;
+--        exponent /= 2;
+--      } else {
+--        exp_result *= exp_base;
+--        --exponent;
+--      }
+--    }
+--    return exp_result;
+--  }
+   function Make_Mod_Expon_Function
+     (Mod_Type : Irep;
+      Exponent_Type : Irep;
+      Source_Location : Source_Ptr)
+   return Irep is
+      Function_Name : constant String := Fresh_Var_Name ("mod_op_expon");
+      Function_Body : constant Irep := Make_Code_Block (Source_Location);
+      Function_Params : constant Irep := Make_Parameter_List;
+      Base_Arg : constant Irep := Create_Fun_Parameter
+        (Fun_Name => Function_Name,
+         Param_Name => "base",
+         Param_Type => Mod_Type,
+         Param_List => Function_Params,
+         A_Symbol_Table => Global_Symbol_Table,
+         Source_Location => Source_Location);
+      Exponent_Arg : constant Irep := Create_Fun_Parameter
+        (Fun_Name => Function_Name,
+         Param_Name => "exponent",
+         Param_Type => Exponent_Type,
+         Param_List => Function_Params,
+         A_Symbol_Table => Global_Symbol_Table,
+         Source_Location => Source_Location);
+
+      Function_Type : constant Irep := Make_Code_Type
+        (Parameters => Function_Params,
+         Ellipsis => False,
+         Return_Type => Mod_Type,
+         Inlined => False,
+         Knr => False);
+
+      Base_Sym : constant Irep := Param_Symbol (Base_Arg);
+      Exponent_Sym : constant Irep := Param_Symbol (Exponent_Arg);
+
+      Expon_Result : constant Irep := Fresh_Var_Symbol_Expr
+        (Ty => Mod_Type,
+         Infix => "mod_expon_result");
+      Declare_Expon_Result : constant Irep := Make_Code_Decl
+        (Symbol => Expon_Result,
+         Source_Location => Source_Location);
+      Expon_Divisible_By_Two : constant Irep := Make_Op_Eq
+        (Lhs => Make_Op_Mod
+           (Lhs => Exponent_Sym,
+            Rhs => Integer_Constant_To_Expr
```

Comment:
This and other integer constants are used multiple times: could be made shorter if you stored and reused their `irep`.

Reply:
I considered it, but I'm not sure if it's enough of an improvement to be worth doing.